### ISSUE TITLE
[BUG] Nothing works as expected really :)

### ISSUE THREAD

**Issue by eosfor** [Azure/PSRule.Rules.Azure#3298]
### Description of the issue

Dear @BernieWhite 

Sorry to bother you with this, but it feels like either I di not understand something or something is not working. I've just spent a few hours playing with it and was not able to make it work at all. Please, help me 🥺 

The source is in [here](https://github.com/eosfor/scripting-notes/tree/psrule-test/scripts/psrule-demo). You can just spin up a dev container in a code space and see everything. Tests are under `scripts/psrule-demo`

[![codespaces](https://github.com/codespaces/badge.svg)](https://github.com/codespaces/new?repo=eosfor/scripting-notes&ref=psrule-test).

This is a folder structure

```console
PS /workspaces/scripting-notes> cd ./scripts/
PS /workspaces/scripting-notes/scripts> tree
.
└── psrule-demo
    ├── deployments
    │   └── non-prod
    │       └── main.bicep
    ├── modules
    │   ├── appserviceplan.bicep
    │   ├── functionapp.bicep
    │   ├── network.bicep
    │   ├── servicebus.bicep
    │   └── subnets.bicep
    ├── ps-rule.yaml
    └── readme.md

5 directories, 8 files
PS /workspaces/scripting-notes/scripts> 
```

This is a list of things that I tried.

# State of the art

The powershell command below

```powershell
cd /workspaces/scripting-notes/scripts/psrule-demo/
Invoke-PSRule -Option ./ps-rule.yaml -InputObject . -Path ./.ps-rule -Verbose -Debug
```

return

```console
Invoke-PSRule: Cannot process argument transformation on parameter 'Option'. Cannot convert value "./ps-rule.yaml" to type "PSRule.Configuration.PSRuleOption". Error: "Options file does not exist."
```

But i would expect it tu run successfully.

Another attempt, with full paths

```powershell
Invoke-PSRule -Option /workspaces/scripting-notes/scripts/psrule-demo/ps-rule.yaml -InputObject /workspaces/scripting-notes/scripts/psrule-demo -Path /workspaces/scripting-notes/scripts/psrule-demo/.ps-rule -Verbose -Debug
```

return

```console
VERBOSE: [Invoke-PSRule] BEGIN::
VERBOSE: [New-PSRuleOption] BEGIN::
VERBOSE: [New-PSRuleOption] END::
VERBOSE: [PSRule][D] -- Scanning for source files: /workspaces/scripting-notes/scripts/psrule-demo/.ps-rule
VERBOSE: [PSRule][D] -- Scanning for source files: .ps-rule/
VERBOSE: [PSRule][D] -- Discovering rules in: /workspaces/scripting-notes/scripts/psrule-demo/.ps-rule/local.ArchitectureVerification.Rule.ps1
VERBOSE: [PSRule][D] -- Found .\local.Architecture.Verification in /workspaces/scripting-notes/scripts/psrule-demo/.ps-rule/local.ArchitectureVerification.Rule.ps1
VERBOSE: [PSRule][R][0][.\local.Architecture.Verification] :: db144ec1c65441c05eb49892888c704121ae5c84
DEBUG: Target failed Type precondition
VERBOSE: [Invoke-PSRule] END::
```

But my expectation here that it should find the rule, which it does, then it should run a rule, find a resource of a type `Microsoft.Network/virtualNetworks` which is defined in [modules/network.bicep](/scripts/psrule-demo/modules/network.bicep), and return `success`.

The scenario is simple, there is a, say, Architect, who defines how a certain solution should look in  Azure. Basically in most cases it s a diagram in visio or smth like that. Then an engineer looks at the diagram and tries to implement it in bicep. So now there is a need for an Architect, to verify if whatever has been developed matched the design. Basically it is about checking whether or not all required resources are in their places, connected to correct vnets/subnets with correct private endpoints, etc.

### Reproduction

The source is in [here](https://github.com/eosfor/scripting-notes/tree/psrule-test/scripts/psrule-demo). You can just spin up a dev container in a code space and see everything. Tests are under `scripts/psrule-demo`

[![codespaces](https://github.com/codespaces/badge.svg)](https://github.com/codespaces/new?repo=eosfor/scripting-notes&ref=psrule-test).

### Version of PSRule

2.9.0

**Comment by eosfor** [comment]
Here is the result of it

```powershell
PS /workspaces/scripting-notes/scripts/psrule-demo> Invoke-PSRule -InputObject . -Verbose -Debug
```

results in

```console
VERBOSE: [Invoke-PSRule] BEGIN::
VERBOSE: [New-PSRuleOption] BEGIN::
VERBOSE: Attempting to read: 
VERBOSE: [New-PSRuleOption] END::
VERBOSE: [PSRule][D] -- Scanning for source files: .ps-rule/
VERBOSE: [PSRule][D] -- Discovering rules in: /workspaces/scripting-notes/scripts/psrule-demo/.ps-rule/local.ArchitectureVerification.Rule.ps1
VERBOSE: [PSRule][D] -- Found .\local.Architecture.Verification in /workspaces/scripting-notes/scripts/psrule-demo/.ps-rule/local.ArchitectureVerification.Rule.ps1
VERBOSE: [PSRule][R][0][.\local.Architecture.Verification] :: d606135e18f175bdf2183746f1c94efd0ea6a9e1
DEBUG: Target failed Type precondition
VERBOSE: [Invoke-PSRule] END::
```

However here I would still expect it executing the rule for the nested resource.

**Comment by BernieWhite** [comment]
@eosfor To file the files, PSRule for Azure requires the command line set slightly differently (https://azure.github.io/PSRule.Rules.Azure/creating-your-pipeline/#__tabbed_1_3). 

So, try rewriting your command line to the below because `-InputObject` is intended to be used if you are passing objects in memory to PSRule. PSRule will scan recursively from the path `.` and find files but ignore files excluded by gitignore by default and will ignore files based on the `input.pathIgnore` option.

```powershell
Invoke-PSRule -Format File -InputPath . -Verbose -Debug
```

At that point you will find that files are discovered, and you need to uncomment parts of `ps-rule.yaml` under the `input.ignorePath` because you will get expansion issues.

For example:

```yaml
input:
  fileObjects: true
  pathIgnore:
    # Ignore other files in the repository.
    - '**'

    # Include deployments.
    - '!**/deployments/**/*.bicep'
```

PSRule for Azure does a full analysis so a basic requirement is that all parameters have a value, as a result your files directly under `modules/` are not compatible but under `deployments/` they are. The `input.ignorePath` option above effectively ignores all files but then includes files under a deployments path. 

Testing modules can be achieved by creating a dedicated tests file for unit testing and then updating `input.ignorePath`. See here: https://azure.github.io/PSRule.Rules.Azure/using-bicep/#testing-bicep-modules, and there are examples https://github.com/Azure/PSRule.Rules.Azure-quickstart/blob/main/modules/keyvault/v1/.tests/main.tests.bicep and the AVM repository also has plenty of examples: https://github.com/Azure/bicep-registry-modules/tree/main/avm/res. Did a session on it here: https://youtu.be/AcLNall-p5A?si=mY3vlGy77LatBtvF

---

Also based on your scenario:

> The scenario is simple, there is a, say, Architect, who defines how a certain solution should look in Azure. Basically in most cases it s a diagram in visio or smth like that. Then an engineer looks at the diagram and tries to implement it in bicep. So now there is a need for an Architect, to verify if whatever has been developed matched the design. Basically it is about checking whether or not all required resources are in their places, connected to correct vnets/subnets with correct private endpoints, etc.

I'd suggest flattening the directory structure so that `.ps-rule/` path and `ps-rule.yaml` are in the root of the repository.

Why? This is mostly because the default assumption in PSRule is that it is running from root of the repository as the working path. You can change that, but it will be harder to maintain in the long run, and there are some breaking changes in PSRule v3 (preview) that affect this. Additionally, there is a VSCode extension for PSRule (see https://azure.github.io/PSRule.Rules.Azure/install/) that will make it easier for engineers to edit then test locally before committing changes.

Also there is an sample here that might be of interest as your progress your setup: https://github.com/Azure/PSRule.Rules.Azure/blob/main/samples/rules/BicepModuleRequires/README.md

**Comment by eosfor** [comment]
Thanks for your answers, @BernieWhite ! Here’s what I’ve done based on your advice:

I created a new repository and moved all the templates into the root directory. Now the structure looks like this:

```console
PS /workspaces/psrule-demo> tree -I .git -I .devcontainer -a
.
├── deployments
│   └── non-prod
│       └── main.bicep
├── .gitignore
├── LICENSE
├── modules
│   ├── appserviceplan.bicep
│   ├── functionapp.bicep
│   ├── network.bicep
│   ├── servicebus.bicep
│   └── subnets.bicep
├── .ps-rule
│   └── local.ArchitectureVerification.Rule.ps1
├── ps-rule.yaml
└── README.md

5 directories, 11 files
```

I believe this satisfies the requirement of having a “flat” structure, with the `.ps-rule/` directory and the `ps-rule.yaml` file located in the root of the repository.

Now, I’m able to run the task you suggested. However, when executing the task as follows:

```console
Executing task: Assert-PSRule -Format File -Path './.ps-rule/' -InputPath . -Outcome Fail, Error;

Using PSRule v2.9.0

   WARN  The option 'Execution.NotProcessedWarning' is deprecated and will be removed with PSRule v3. See http://aka.ms/ps-rule/deprecations for more detail.

Rules processed: 0, failed: 0, errored: 0
Run a2f600f5aa9211ff11565fd4e813e0c3ffd28523 completed in 00:00:00.0722787
 *  Terminal will be reused by tasks, press any key to close it.
```

I get a message indicating no rules were processed.

If I attempt to run the rule directly from the command line like this:

```powershell
PS /workspaces/psrule-demo> Invoke-PSRule -InputObject . -Verbose -Debug
```

The rule is discovered but doesn’t execute against the resources. The debug message explicitly states `Target failed Type precondition`:

```console
VERBOSE: [Invoke-PSRule] BEGIN::
VERBOSE: [New-PSRuleOption] BEGIN::
VERBOSE: Attempting to read: 
VERBOSE: [New-PSRuleOption] END::
VERBOSE: [PSRule][D] -- Scanning for source files: .ps-rule/
VERBOSE: [PSRule][D] -- Discovering rules in: /workspaces/psrule-demo/.ps-rule/local.ArchitectureVerification.Rule.ps1
VERBOSE: [PSRule][D] -- Found .\local.Architecture.Verification in /workspaces/psrule-demo/.ps-rule/local.ArchitectureVerification.Rule.ps1
VERBOSE: [PSRule][R][0][.\local.Architecture.Verification] :: d606135e18f175bdf2183746f1c94efd0ea6a9e1
DEBUG: Target failed Type precondition
VERBOSE: [Invoke-PSRule] END::
```

However, the rule explicitly targets Microsoft.Network/virtualNetworks, and such a resource does indeed exist in my templates. After building main.bicep, the resulting JSON file clearly contains a Microsoft.Network/virtualNetworks resource somewhere in its hierarchy.

To clarify my goal further: I want a rule that scans through the repository templates and verifies conditions such as:
	•	Is there a Virtual Network?
	•	Is there an App Service Plan?
	•	Is there a Service Bus?
	•	Is the Service Bus attached to the VNET via a private endpoint?

At the moment, however, the rule isn’t triggered even though the required VNET resource is defined in the Bicep templates.

You can check this directly in my repo using the provided Codespaces link:

[![codespaces](https://github.com/codespaces/badge.svg)](https://github.com/codespaces/new?repo=eosfor/psrule-demo&ref=main).

Additionally, as a user, I would find it very useful if I could:

- Verify that PSRule expands Bicep variables and modules and correctly produces ARM JSON. Ideally, seeing this explicitly logged in the debug or verbose PowerShell output would be very helpful.
- Set breakpoints in `.ps1` rule scripts to see exactly what inputs the rules receive, allowing me to debug the rules step-by-step. If there’s a recommended approach or tool to achieve this, your advice would be greatly appreciated.

**Comment by BernieWhite** [comment]
Thanks @eosfor. That's progress. Can you confirm that you have the PSRule.Rules.Azure PowerShell module installed?

Also, you can either uncomment:

```
# include:
#   module:
#     - PSRule.Rules.Azure
```
 
Or add `-Module PSRule.Rules.Azure` to your Invoke-PSRule command line.

Since PSRule for Azure actually does the expansion, it need to be included.

**Comment by eosfor** [comment]
@BernieWhite 

Added the following 

```console
PS /workspaces/psrule-demo> bicep --version
Bicep CLI version 0.34.44 (1f661a23c4)
PS /workspaces/psrule-demo> Get-Module az -ListAvailable                      

    Directory: /home/vscode/.local/share/powershell/Modules

ModuleType Version    PreRelease Name                                PSEdition ExportedCommands
---------- -------    ---------- ----                                --------- ----------------
Script     13.3.0                Az                                  Core,Desk 
```


The task returns

```console
 *  Executing task: Assert-PSRule -Format File -Path './.ps-rule/' -InputPath . -Outcome Fail, Error; 

WARNING: To use PSRule for Azure export cmdlets please install Az.Resources >= 6.7.0 and < 7.0.0. To suppress this warning set the environment variable 'PSRULE_AZURE_RESOURCE_MODULE_NOWARN' to 'true'.
Using PSRule v2.9.0
Using PSRule.Rules.Azure v1.41.4

   WARN  The option 'Execution.NotProcessedWarning' is deprecated and will be removed with PSRule v3. See http://aka.ms/ps-rule/deprecations for more detail.

Rules processed: 0, failed: 0, errored: 0
Run 9b5f9ec8876356852ea95c23169b7b88598da88b completed in 00:00:01.7747520
 *  Terminal will be reused by tasks, press any key to close it. 
```

Running commands manually says

```console
PS /workspaces/psrule-demo> Invoke-PSRule -InputObject .
PS /workspaces/psrule-demo> Assert-PSRule -InputObject .
    ____  _____ ____        __
   / __ \/ ___// __ \__  __/ /__
  / /_/ /\__ \/ /_/ / / / / / _ \
 / ____/___/ / _, _/ /_/ / /  __/
/_/    /____/_/ |_|\__,_/_/\___/

Using PSRule v2.9.0
Using PSRule.Rules.Azure v1.41.4

----------------------------
Explore documentation: https://aka.ms/ps-rule
Contribute and find source: https://github.com/microsoft/PSRule
Report issues: https://github.com/microsoft/PSRule/issues
PSRule.Rules.Azure: https://aka.ms/ps-rule-azure
----------------------------

From repository: https://github.com/eosfor/psrule-demo
  on : main
  at : 9207bf914f8bd65ff7f9ce53027fdadb084b913d


Rules processed: 0, failed: 0, errored: 0
```

However with `debug` option I see that it is now sees various Azure rules. But still skips mine, I guess

```
....
DEBUG: Target failed Type precondition
VERBOSE: [PSRule][R][0][PSRule.Rules.Azure\Azure.BV.Immutable] :: d606135e18f175bdf2183746f1c94efd0ea6a9e1
DEBUG: [PSRule][R][Trace] -- PSRule.Rules.Azure\Azure.BV.Immutable
DEBUG: Target failed Type precondition
VERBOSE: [PSRule][R][0][PSRule.Rules.Azure\Azure.ASG.Name] :: d606135e18f175bdf2183746f1c94efd0ea6a9e1
DEBUG: [PSRule][R][Trace] -- PSRule.Rules.Azure\Azure.ASG.Name
DEBUG: Target failed Type precondition
VERBOSE: [PSRule][R][0][PSRule.Rules.Azure\Azure.Redis.MinSKU] :: d606135e18f175bdf2183746f1c94efd0ea6a9e1
DEBUG: [PSRule][R][Trace] -- PSRule.Rules.Azure\Azure.Redis.MinSKU
DEBUG: Target failed Type precondition
VERBOSE: [PSRule][R][0][PSRule.Rules.Azure\Azure.Redis.NonSslPort] :: d606135e18f175bdf2183746f1c94efd0ea6a9e1
DEBUG: [PSRule][R][Trace] -- PSRule.Rules.Azure\Azure.Redis.NonSslPort
DEBUG: Target failed Type precondition
VERBOSE: [PSRule][R][0][PSRule.Rules.Azure\Azure.Redis.MinTLS] :: d606135e18f175bdf2183746f1c94efd0ea6a9e1
DEBUG: [PSRule][R][Trace] -- PSRule.Rules.Azure\Azure.Redis.MinTLS
DEBUG: Target failed Type precondition
VERBOSE: [PSRule][R][0][PSRule.Rules.Azure\Azure.Redis.PublicNetworkAccess] :: d606135e18f175bdf2183746f1c94efd0ea6a9e1
DEBUG: [PSRule][R][Trace] -- PSRule.Rules.Azure\Azure.Redis.PublicNetworkAccess
DEBUG: Target failed Type precondition
VERBOSE: [PSRule][R][0][PSRule.Rules.Azure\Azure.Redis.EntraID] :: d606135e18f175bdf2183746f1c94efd0ea6a9e1
DEBUG: [PSRule][R][Trace] -- PSRule.Rules.Azure\Azure.Redis.EntraID
DEBUG: Target failed Type precondition
VERBOSE: [Invoke-PSRule] END::
```

**Comment by BernieWhite** [comment]
See my branch here: https://github.com/BernieWhite/scripting-notes-tests/tree/psrule-test

Run with the command ` Assert-PSRule -Format File -InputPath .`

Get the output:

```text
    ____  _____ ____        __
   / __ \/ ___// __ \__  __/ /__
  / /_/ /\__ \/ /_/ / / / / / _ \
 / ____/___/ / _, _/ /_/ / /  __/
/_/    /____/_/ |_|\__,_/_/\___/

Using PSRule v2.9.0
Using PSRule.Rules.Azure v1.41.2

----------------------------
Explore documentation: https://aka.ms/ps-rule
Contribute and find source: https://github.com/microsoft/PSRule
Report issues: https://github.com/microsoft/PSRule/issues
PSRule.Rules.Azure: https://aka.ms/ps-rule-azure
----------------------------

> D:\Dev\repos\scripting-notes-tests\deployments\non-prod\main.bicep : Microsoft.Resources/deployments [7/7]

   PASS  Azure.Deployment.OutputSecretValue (AZR-000279)
   PASS  Azure.Deployment.AdminUsername (AZR-000284)
   PASS  Azure.Deployment.SecureParameter (AZR-000408)
   PASS  Azure.Deployment.SecureValue (AZR-000316)
   PASS  Azure.Deployment.OuterSecret (AZR-000331)
   PASS  Azure.Template.ParameterStrongType (AZR-000227)
   PASS  Azure.Template.ExpressionLength (AZR-000228)

> vnetDeployment : Microsoft.Resources/deployments [7/7]

   PASS  Azure.Deployment.OutputSecretValue (AZR-000279)
   PASS  Azure.Deployment.AdminUsername (AZR-000284)
   PASS  Azure.Deployment.SecureParameter (AZR-000408)
   PASS  Azure.Deployment.SecureValue (AZR-000316)
   PASS  Azure.Template.ParameterStrongType (AZR-000227)
   PASS  Azure.Template.ExpressionLength (AZR-000228)
   PASS  Azure.Deployment.Name (AZR-000359)

> my-vnet : Microsoft.Network/virtualNetworks [5/7]

   FAIL  Azure.Resource.UseTags (AZR-000166)

  Azure resources should be tagged using a standard convention.

  Template: deployments/non-prod/main.bicep:60:13

  Recommend:
  Consider tagging resources using a standard convention. Identify mandatory and
  optional tags then tag all resources and resource groups using this standard.
  Also consider using Azure Policy to enforce mandatory tags.

  Reason:
  - The resource is not tagged.
  - Path tags: Does not exist.

  Help:
  - https://azure.github.io/PSRule.Rules.Azure/en/rules/Azure.Resource.UseTags/

   PASS  Azure.Resource.AllowedRegions (AZR-000167)
   PASS  Azure.VNET.UseNSGs (AZR-000263)
   PASS  Azure.VNET.SingleDNS (AZR-000264)
   PASS  Azure.VNET.SubnetName (AZR-000267)
   FAIL  local.Architecture.Verification

  Template: deployments/non-prod/main.bicep:60:13

  Reason:
  - Path name: Is set to 'my-vnet'.

   PASS  Azure.VNET.Name (AZR-000268)

> subnetsDeployment : Microsoft.Resources/deployments [7/7]

   PASS  Azure.Deployment.OutputSecretValue (AZR-000279)
   PASS  Azure.Deployment.AdminUsername (AZR-000284)
   PASS  Azure.Deployment.SecureParameter (AZR-000408)
   PASS  Azure.Deployment.SecureValue (AZR-000316)
   PASS  Azure.Template.ParameterStrongType (AZR-000227)
   PASS  Azure.Template.ExpressionLength (AZR-000228)
   PASS  Azure.Deployment.Name (AZR-000359)

> planDeployment : Microsoft.Resources/deployments [7/7]

   PASS  Azure.Deployment.OutputSecretValue (AZR-000279)
   PASS  Azure.Deployment.AdminUsername (AZR-000284)
   PASS  Azure.Deployment.SecureParameter (AZR-000408)
   PASS  Azure.Deployment.SecureValue (AZR-000316)
   PASS  Azure.Template.ParameterStrongType (AZR-000227)
   PASS  Azure.Template.ExpressionLength (AZR-000228)
   PASS  Azure.Deployment.Name (AZR-000359)

> myfunc5f3e65afb63bb-plan : Microsoft.Web/serverfarms [1/3]

   FAIL  Azure.AppService.AvailabilityZone (AZR-000442)

  Deploy app service plan instances using availability zones in supported regions to ensure high availability and resilience.

  Template: deployments/non-prod/main.bicep:197:13

  Recommend:
  Consider using enabling zone redundancy using availability zones to improve the
  resiliency of your solution.

  Reason:
  - Path sku.name: The app service plan (myfunc5f3e65afb63bb-plan) is not deployed with a SKU that supports zone-redundancy.
  - Path sku.tier: The app service plan (myfunc5f3e65afb63bb-plan) is not deployed with a SKU that supports zone-redundancy.
  - Path properties.zoneRedundant: Does not exist.

  Help:
  - https://azure.github.io/PSRule.Rules.Azure/en/rules/Azure.AppService.AvailabilityZone/

   FAIL  Azure.Resource.UseTags (AZR-000166)

  Azure resources should be tagged using a standard convention.

  Template: deployments/non-prod/main.bicep:197:13

  Recommend:
  Consider tagging resources using a standard convention. Identify mandatory and
  optional tags then tag all resources and resource groups using this standard.
  Also consider using Azure Policy to enforce mandatory tags.

  Reason:
  - The resource is not tagged.
  - Path tags: Does not exist.

  Help:
  - https://azure.github.io/PSRule.Rules.Azure/en/rules/Azure.Resource.UseTags/

   PASS  Azure.Resource.AllowedRegions (AZR-000167)

> sbDeployment : Microsoft.Resources/deployments [7/7]

   PASS  Azure.Deployment.OutputSecretValue (AZR-000279)
   PASS  Azure.Deployment.AdminUsername (AZR-000284)
   PASS  Azure.Deployment.SecureParameter (AZR-000408)
   PASS  Azure.Deployment.SecureValue (AZR-000316)
   PASS  Azure.Template.ParameterStrongType (AZR-000227)
   PASS  Azure.Template.ExpressionLength (AZR-000228)
   PASS  Azure.Deployment.Name (AZR-000359)

> sb5f3e65afb63bb : Microsoft.ServiceBus/namespaces [1/4]

   FAIL  Azure.Resource.UseTags (AZR-000166)

  Azure resources should be tagged using a standard convention.

  Template: deployments/non-prod/main.bicep:254:13

  Recommend:
  Consider tagging resources using a standard convention. Identify mandatory and
  optional tags then tag all resources and resource groups using this standard.
  Also consider using Azure Policy to enforce mandatory tags.

  Reason:
  - The resource is not tagged.
  - Path tags: Does not exist.

  Help:
  - https://azure.github.io/PSRule.Rules.Azure/en/rules/Azure.Resource.UseTags/

   PASS  Azure.Resource.AllowedRegions (AZR-000167)
   FAIL  Azure.ServiceBus.DisableLocalAuth (AZR-000178)

  Authenticate Service Bus publishers and consumers with Entra ID identities.

  Template: deployments/non-prod/main.bicep:254:13

  Recommend:
  Consider only using Entra ID identities to publish or consume messages from
  Service Bus. Then disable authentication based on access keys or SAS tokens.

  Reason:
  - Path properties.disableLocalAuth: The field 'properties.disableLocalAuth' does not exist.

  Help:
  - https://azure.github.io/PSRule.Rules.Azure/en/rules/Azure.ServiceBus.DisableLocalAuth/

   FAIL  Azure.ServiceBus.MinTLS (AZR-000315)

  Service Bus namespaces should reject TLS versions older than 1.2.

  Template: deployments/non-prod/main.bicep:254:13

  Recommend:
  Consider configuring the minimum supported TLS version for Service Bus clients
  to be 1.2. Also consider enforcing this setting using Azure Policy.

  Reason:
  - Path properties.minimumTlsVersion: The field 'properties.minimumTlsVersion' does not exist.

  Help:
  - https://azure.github.io/PSRule.Rules.Azure/en/rules/Azure.ServiceBus.MinTLS/

> functionDeployment : Microsoft.Resources/deployments [7/7]

   PASS  Azure.Deployment.OutputSecretValue (AZR-000279)
   PASS  Azure.Deployment.AdminUsername (AZR-000284)
   PASS  Azure.Deployment.SecureParameter (AZR-000408)
   PASS  Azure.Deployment.SecureValue (AZR-000316)
   PASS  Azure.Template.ParameterStrongType (AZR-000227)
   PASS  Azure.Template.ExpressionLength (AZR-000228)
   PASS  Azure.Deployment.Name (AZR-000359)

> myfunc5f3e65afb63bb : Microsoft.Web/sites [6/11]

   FAIL  Azure.AppService.MinTLS (AZR-000073)

  App Service should not accept weak or deprecated transport protocols for client-server communication.

  Template: deployments/non-prod/main.bicep:328:13

  Recommend:
  Consider configuring the minimum supported TLS version to be 1.2. Also consider
  using Azure Policy to audit or enforce this configuration.

  Reason:
  - Path properties.siteConfig.minTlsVersion: Minimum TLS version is set to {0}.

  Help:
  - https://azure.github.io/PSRule.Rules.Azure/en/rules/Azure.AppService.MinTLS/

   PASS  Azure.AppService.RemoteDebug (AZR-000074)
   PASS  Azure.AppService.NETVersion (AZR-000075)
   PASS  Azure.AppService.PHPVersion (AZR-000076)
   FAIL  Azure.AppService.HTTP2 (AZR-000078)

  Use HTTP/2 instead of HTTP/1.x to improve protocol efficiency.

  Template: deployments/non-prod/main.bicep:328:13

  Recommend:
  Consider using HTTP/2 for Azure Services apps to improve protocol efficiency.

  Reason:
  - Path Properties.siteConfig.http20Enabled: Does not exist.

  Help:
  - https://azure.github.io/PSRule.Rules.Azure/en/rules/Azure.AppService.HTTP2/

   PASS  Azure.AppService.NodeJsVersion (AZR-000428)
   FAIL  Azure.Resource.UseTags (AZR-000166)

  Azure resources should be tagged using a standard convention.

  Template: deployments/non-prod/main.bicep:328:13

  Recommend:
  Consider tagging resources using a standard convention. Identify mandatory and
  optional tags then tag all resources and resource groups using this standard.
  Also consider using Azure Policy to enforce mandatory tags.

  Reason:
  - The resource is not tagged.
  - Path tags: Does not exist.

  Help:
  - https://azure.github.io/PSRule.Rules.Azure/en/rules/Azure.Resource.UseTags/

   PASS  Azure.Resource.AllowedRegions (AZR-000167)
   FAIL  Azure.AppService.ManagedIdentity (AZR-000082)

  Configure managed identities to access Azure resources.

  Template: deployments/non-prod/main.bicep:328:13

  Recommend:
  Consider configuring a managed identity for each App Service app. Also consider
  using managed identities to authenticate to related Azure services.

  Reason:
  - Path identity.type: The field 'identity.type' does not exist.

  Help:
  - https://azure.github.io/PSRule.Rules.Azure/en/rules/Azure.AppService.ManagedIdentity/

   FAIL  Azure.AppService.ARRAffinity (AZR-000083)

  Disable client affinity for stateless services.

  Template: deployments/non-prod/main.bicep:328:13

  Recommend:
  Azure App Service sites make use of Application Request Routing (ARR) by
  default. Consider disabling ARR affinity for stateless applications.

  Reason:
  - Path properties.clientAffinityEnabled: The field 'properties.clientAffinityEnabled' does not exist.

  Help:
  - https://azure.github.io/PSRule.Rules.Azure/en/rules/Azure.AppService.ARRAffinity/

   PASS  Azure.AppService.UseHTTPS (AZR-000084)

> function-pe : Microsoft.Network/privateEndpoints [2/3]

   FAIL  Azure.Resource.UseTags (AZR-000166)

  Azure resources should be tagged using a standard convention.

  Template: deployments/non-prod/main.bicep:353:13

  Recommend:
  Consider tagging resources using a standard convention. Identify mandatory and
  optional tags then tag all resources and resource groups using this standard.
  Also consider using Azure Policy to enforce mandatory tags.

  Reason:
  - The resource is not tagged.
  - Path tags: Does not exist.

  Help:
  - https://azure.github.io/PSRule.Rules.Azure/en/rules/Azure.Resource.UseTags/

   PASS  Azure.Resource.AllowedRegions (AZR-000167)
   PASS  Azure.PrivateEndpoint.Name (AZR-000153)

Rules processed: 70, failed: 13, errored: 0
Run c5b79928568b3e725a4acd4095bf6261139f598a completed in 00:00:19.1681202
Assert-PSRule: One or more rules reported failure.
```

Note your rule is called one time and returned fail.

**Comment by BernieWhite** [comment]
Key points in `ps-rule.yaml`

```yaml
include:
  module:
    - PSRule.Rules.Azure
```

```yaml
input:
  fileObjects: true
  pathIgnore:
    # Ignore other files in the repository.
    - '**'

    # Include deployments.
    - '!deployments/**/*.bicep'

    # Include module tests.
    # - '!modules/**/*.bicep'
```

```yaml
# Configure binding for local rules.
binding:
  preferTargetInfo: true
  targetType:
    - type
    - resourceType
```

```yaml
configuration:
  # Enable automatic expansion of Azure Bicep source files.
  AZURE_BICEP_FILE_EXPANSION: true
```

**Comment by eosfor** [comment]
Now I see the difference, @BernieWhite , that you so much! The issue was in this little forward slash 😄 

![Image](https://github.com/user-attachments/assets/da19ee55-a429-477e-94f1-12c71680258f)

Now I can continue exploring it, thank you so much! 🙇 

A few more things, If I may:

- How can I troubleshoot these rule `ps1` scripts. Is there any way I can set breakpoints?
- Even thou I included `PSRule.Rules.Azure` module, I do not want it to run in this particular case. How do I make it so?

**Comment by BernieWhite** [comment]
Hi @eosfor, no worries.

PSRule doesn't have a debugger experience today, however it supports standard PowerShell cmdlets in `.Rule.ps1` files, for example `Write-Debug` and `Write-Verbose` will show output when `-Verbose` or `-Debug` is specified.

For simply testing a rule you can run a single rule by specifying the `-Name` argument to only run that rule.

For more broadly only running rules for your organization, then you could do that by configuring the `rule` option. https://microsoft.github.io/PSRule/v2/concepts/PSRule/en-US/about_PSRule_Options/#ruleinclude

For example:

```yaml
rule:
  include:
    - 'local.*'
```

**Comment by BernieWhite** [comment]
@eosfor I am going to close this issue because it appears there is no issue.

Feel free to ask any questions in the future, however we ask they are posted under discussions, because it helps others in the community find the answers to similar questions. Thanks.

**Comment by eosfor** [comment]
Thank you, @BernieWhite
